Expand Section Links documentation and add a section for Custom Anchors (#33531)
github/docs: content/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax.md

{% data reusables.repositories.section-links %}

If you need to determine the anchor for a heading in a file you are editing, you can use the following basic rules:

* Letters are converted to lower-case.

> * Spaces are replaced by hyphens (`-`). Any other whitespace characters are removed.

Reviewer: This now misses any info about punctuation. (Which is perhaps better, as the rules are somewhat wild…)

I think `.;+()` is removed, but the last time i checked `?` not etc. — is there a rule for these characters?

Reviewer: ```suggestion
* Spaces are replaced by hyphens (`-`). Any other whitespace or punctuation characters are removed.
```

* Leading and trailing whitespace are removed.

> * Paired formatting markup characters are removed, leaving only their contents (for example, `_italics_` becomes `italics`).

Reviewer: ```suggestion
* Markup formatting is removed, leaving only the contents (for example, `_italics_` becomes `italics`).
```
I'm thinking about simplifying this in a way that not only Markdown formatting is removed, but basically any HTML formatting is removed … e.g. if the heading has a `<mark> foo </mark>` these would also not be part of the fragment I suppose…

* If the automatically generated anchor for a heading is identical to an earlier anchor in the same document, a unique identifier is generated by appending a hyphen and an auto-incrementing integer.

For more detailed information on the requirements of URI fragments, see [RFC 3986: Uniform Resource Identifier (URI): Generic Syntax, Section 3.5](https://www.rfc-editor.org/rfc/rfc3986#section-3.5).

The code block below demonstrates the basic rules used to generate anchors from headings in rendered content.

```markdown
# Example headings 

## Sample Section

## This'll  be a _Helpful_ Section About the Greek Letter Θ!

> A heading containing invalid characters, UTF-8 characters, two consecutive spaces between the first and second words, and formatting.

Reviewer: ```suggestion
A heading containing characters not allowed in fragments, UTF-8 characters, two consecutive spaces between the first and second words, and formatting.
```

## This heading is not unique in the file

TEXT 1

## This heading is not unique in the file

TEXT 2

# Links to the example headings above

Link to the sample section: [Link Text](#sample-section).

Link to the helpful section: [Link Text](#thisll--be-a-helpful-section-about-the-greek-letter-Θ).

Link to the first non-unique section: [Link Text](#this-heading-is-not-unique-in-the-file).

Link to the second non-unique section: [Link Text](#this-heading-is-not-unique-in-the-file-1).
```

> [!NOTE]
> If you edit a heading, or if you change the order of headings with "identical" anchors, you will also need to update any links to those headings as the anchors will change.

## Relative links

{% data reusables.repositories.relative-links %}

## Custom anchors

You can use standard HTML anchor tags (`<a name="unique-anchor-name"></a>`) to create navigation anchor points for any location in the document. To avoid ambiguous references, use a unique naming scheme for anchor tags, such as adding a prefix to the `name` attribute value.

> [!NOTE]
> Custom anchors will not be included in the document outline/Table of Contents.

> You can link to a custom anchor with the name, `name`, you gave the anchor. The syntax is exactly the same as when you link to an anchor that is automatically generated for a heading.

Reviewer: ```suggestion
You can link to a custom anchor using the value of `name` attribute you gave the anchor. The syntax is exactly the same as when you link to an anchor that is automatically generated for a heading.
```

Reviewer 2: Thanks for the suggestion. Making a minor edit:

```suggestion
You can link to a custom anchor using the value of the `name` attribute you gave the anchor. The syntax is exactly the same as when you link to an anchor that is automatically generated for a heading.
```

For example:

```markdown
# Section Heading

Some body text of this section.

<a name="my-custom-anchor-point"></a>
Some text I want to provide a direct link to, but which doesn't have its own heading.

> ...more content...

Reviewer: ```suggestion
(… more content…)
```

(basically just better ellipsis)

{% data reusables.repositories.section-links %}

If you need to determine the anchor for a heading in a file you are editing, you can use the following basic rules:

* Letters are converted to lower-case.
* Spaces are replaced by hyphens (`-`). Any other whitespace characters are removed.
* Leading and trailing whitespace are removed.
* Paired formatting markup characters are removed, leaving only their contents (for example, `_italics_` becomes `italics`).
* If the automatically generated anchor for a heading is identical to an earlier anchor in the same document, a unique identifier is generated by appending a hyphen and an auto-incrementing integer.

> For more detailed information on the requirements of URI fragments, see [RFC 3986: Uniform Resource Identifier (URI): Generic Syntax, Section 3.5](https://www.rfc-editor.org/rfc/rfc3986#section-3.5).

Reviewer: I have a feeling this RFC is now linked twice from somewhere, TBC / double checking…

The code block below demonstrates the basic rules used to generate anchors from headings in rendered content.

```markdown
# Example headings 

## Sample Section

## This'll  be a _Helpful_ Section About the Greek Letter Θ!
A heading containing invalid characters, UTF-8 characters, two consecutive spaces between the first and second words, and formatting.

## This heading is not unique in the file

TEXT 1

## This heading is not unique in the file

TEXT 2

# Links to the example headings above

Link to the sample section: [Link Text](#sample-section).

Link to the helpful section: [Link Text](#thisll--be-a-helpful-section-about-the-greek-letter-Θ).

Link to the first non-unique section: [Name](#this-heading-is-not-unique-in-the-file).

Link to the second non-unique section: [Name](#this-heading-is-not-unique-in-the-file-1).

> ```

Reviewer: "heading containing invalid characters" sounds a bit weird, the `'` apostrophe is definitely not invalid, (:) just not allowed in the URI fragments — is there a better expression for that?

> [!NOTE]
> If you edit a heading, or if you change the order of headings with "identical" anchors, you will also need to update any links to those headings as the anchors will change.

## Relative links

{% data reusables.repositories.relative-links %}

## Custom anchors

> You can use standard HTML anchor tags (`<a name="unique-anchor-name"></a>`) to create navigation anchor points for any location in the document.

Reviewer 2: ```suggestion
You can use standard HTML anchor tags (`<a name="unique-anchor-name"></a>`) to create navigation anchor points for any location in the document. To avoid ambiguous references, use a unique naming scheme for anchor tags, such as adding a prefix to the `name` attribute value.
```

{% data reusables.repositories.relative-links %}

## Custom anchors

{% data reusables.repositories.custom-anchors %}

> 

Author: I would say no. However, that's how all of those sections seem to be formed, so I was just following suit. 🤷‍♂️

> 

Reviewer: Q: does it need a separate section, or having that contextually appended to some of the existing content would provide more utility and be included with such reusable to all the places they're being currently included?

> 

Reviewer 2: Thanks for the discussion @janbrasna 👋🏻 

> does only one use/occurrence warrant a reusable?

Good point. We usually add a reusable file only if we want to use one or more sentences in more than one location.
